README
[![Build Status](https://travis-ci.org/fmaguire/Bridger_Assembler.svg?branch=master)](https://travis-ci.org/fmaguire/Bridger_Assembler)

================
 Description
================
                                                                                  
 Bridger is an efficient de novo trascriptome assembler for RNA-Seq data. It can assemble 
 all transcripts from short reads (single or paired) without using a reference.  
 The software expects as input RNA-Seq reads in fasta or fastq format, and ouput all
 assembled candidate transcripts in fasta format. Briefly, it works in two step: first,
 Bridger partitions the sequence data into many individual splicing graphs, each capturing 
 the full transcriptional complexity at a given gene or no more than a few genes. Then, 
 Bridger uses a rigorous mathematical model called minimum path cover to search minimal 
 set of paths(transcripts) that can be supported by our data and could explain all observed 
 splicing events of each locus.

 This software is free to use, modify, redistribute without any restrictions, 
 except including the license provided with the distribution. 


================
 Installation
================


 1. Installing Boost

    a) download latest boost and unpack it.

       $ tar zxvf boost_1_47_0.tar.gz

    b) change to the boost directory and run ./bootstrap.sh.

       $ cd  boost_1_47_0
       $ ./bootstrap.sh

    c) Run ./b2 and set the install path. 

       $ ./b2 install --prefix=<YOUR_BOOST_INSTALL_DIRECTORY>

       For example,if you want install boost in /home/czheng/local/boost,the commnd is :
       $ ./b2 install --prefix=/home/czheng/local/boost
     
       If the boost is installed successfully, you would fild two sub-directories in /home/czheng/local/boost/:
       /home/czheng/local/boost/include/ 
       /home/czheng/local/boost/lib/

      Note: The default Boost installation directory is /usr/local. Take note of the boost installation
       directory, beacuase you need to tell the Bridger installer where to find boost later on.

    d) Set the LD_LIBRARY_PATH enviroment variable:
       
       The ~/.bash_profile ($HOME/.bash_profile) or ~/.profile file is executed when you login using console or remotely using ssh. 
       Append the following command to  ~/.bash_profile or ~/.profile file:
       $ export LD_LIBRARY_PATH=/home/czheng/local/boost/lib:$LD_LIBRARY_PATH

       Save and close the file.

       OR
       
       just type the command:
       $ export LD_LIBRARY_PATH=/home/czheng/local/boost/lib:$LD_LIBRARY_PATH  

       Note: please replace "/home/czheng/local/boost/lib" with your own directory "<YOUR_BOOST_INSTALL_DIRECTORY>/lib"
       If you do not set this variable , you would possible see the follwoing error information:
        "error while loading shared libraries: libboost_serialization.so.1.47.0: cannot open shared object file: No such file or directory"

 2. Building Bridger [Make sure Boost has been installed successfully]
  
    a) Unpack the Bridger and change to the Bridger direcotry.

       $ tar zxvf Bridger_r2013-06-02.tar.gz
       $ cd Bridger_r2013-06-02

    b) Configure Bridger. If Boost is installed somewhere other than /usr/local, you will need to tell
       the installer where to find it using --with-boost option.

       $ ./configure --with-boost=/home/czheng/local/boost/
       Note: please replace "/home/czheng/local/boost/" with your own directory "<YOUR_BOOST_INSTALL_DIRECTORY>"
     

    c) Make Bridger.

       $ make
      
      note: If you build boost suffessfully without using --prefix option, the following commands may need before "make":
          export LIBS="-L/home/czheng/boost_1_47_0/stage/lib" (replace "/home/czheng/boost_1_47_0/" with your own directory)
          export CPPFLAGS="-I/home/czheng/boost_1_47_0/"

 3. Test the installation. Test data are provided with sofeware distribution in the sample_test directory.
     $ cd src
     $ ./Assemble -h
     you would see:
     ===============================================================================
     Usage: Assemble [--reads/--kmers] <filename>  [opts] 
     ===============================================================================
     **Required :
     --reads/-i <string>           : the name of the file containing reads

     ** Optional :
     --kmer_length/-k <int>        : length of kmer, default: 25.
     --double_stranded_mode        : set it true if double stranded mode.
     --fr_strand<int>              : strand specific protocol, default: 1 
                                       ( 1 : fr-firststrand, e.g. dUTP, NSR, NNSR 
                                         2 : fr-secondstrand, e.g. Strandard SOLID ) 
     --paired_end                  : set it true if paired reads.
     --min_seed_coverage <int>     : minimum coverage of seed kmer, default: 2.
     --min_seed_entropy <float>    : minimum entropy of seed kmer, default: 1.5
     --min_kmer_coverage <int>     : minimum coverage of kmer used to extend, default: 1.
     --min_kmer_entropy <float>    : minimum entroy of kmer used to extend, default: 0.0
     --min_junction_coverage <int> : minimum of the coverage of a junction, default: 2.
     --min_ratio_non_error <float> : min ratio for low/high alternative extension that is 
                                          not an error, default: 0.05.
     --pair_gap_length             : gap length of paired reads, default: 200.
     --out_dir/-o <string>         : name of directory for output, default : ./RawGraphs 
     --help/-h                     : display the help information.

    ===============================================================================
     

     Note : If you  see the error information like "error while loading shared libraries: libboost_serialization.so.1.47.0: 
               cannot open shared object file: No such file or directory", please set the LD_LIBRARY_PATH variable by command:
              "export LD_LIBRARY_PATH=/home/czheng/local/boost_1_47_0/lib:$LD_LIBRARY_PATH"
 
     Test Bridger with a small data:

     $ cd sample_test/
     $ ./run_Me.sh



===============
 Uasge
===============

 A typical command maybe like this:
 
 ./Bridger.pl --seqType fq --left reads.left.fq --right reads.right.fq --CPU 6
 
 Note:
   "--SS_lib_type" is recommended to be used for strand-specific RNA-Seq data.
 
  For more inforamtion, use --help option. 
  or, visit : http://BridgerRNASeq.sourceforge.net


================
 Changelog
================

 Version r2014-12-01
 - latest version


===============
 Authors
===============

 Zheng Chang designed and wrote Bridger, with substantial technical input
 from Yu Zhang and Cody Ashby.  


================
 Contact
================

 Any questions, problems, bugs are welcome and should be dumped to
 Zheng Chang <changzmaths@gmail.com>

 Created on June 25, 2012.
 


ChangeLog
===========
2013-06-02
===========

plugins:
    include a software called RSEM to estimate abundance of assembled assemblies.

perlLib:
    add a perl script called run_RSEM_align_n_estimate.pl simply maps the familiar parameters 
    used in assembley to those of the RSEM software and then executes RSEM accordingly.

plugins:
    add two scripts:
    *run_abundance_estimation.sh : test abundance estimation using RSEM software.
    *clean.pl : clean those files generated during test Bridger and RSEM.


src/assemble.cpp
/*
 *
 * Copyright(c) 2012,2013. Zheng Chang.
 *
 * This file is part of program Bridger, which is used to de novo
 * assemble transcriptome from RNASeq data.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *
 * Problems, suggestions, comments, etc ? Contact :
 *    Zheng Chang  <changzmaths@gmail.com>
 *
 * Created on June 25, 2012.
 */

#include "loadreads.h"
#include "common.h"
#include "splicing_graph.h"
#include "kmerhash.h"
#include "assemble.h"

#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <fstream>
#include <stdlib.h>
#include <errno.h>
#include <ctime>

#include <boost/filesystem.hpp>


int parse_options(int argc, char* argv[]) {

  int option_index = 0;
  int next_option;

  do {
    next_option = getopt_long(argc, argv, short_options, long_options, &option_index);
    switch (next_option) {
    case -1:     /* Done with options. */
      break;
    case 'k':
      g_kmer_length = atoi(optarg);
      break;
    case 'i':
      reads_file = optarg;
      break;
    case 'o':
      out_dir = optarg;
      break;
    case 'h':
      g_help = true;
      break;
    case 'p':
      CPU = atoi(optarg);
      break;
    case OPT_KMERS:
      kmers_file = optarg;
      break;
    case OPT_DOUBLE_STRANDED_MODE:
      g_double_stranded_mode = true;
      break;
    case OPT_IS_PAIR_END:
      g_is_paired_end = true;
      break;
    case OPT_MIN_KMER_COVERAGE:
      g_min_kmer_coverage = atoi(optarg);
      break;
    case OPT_MIN_KMER_ENTROPY:
      g_min_kmer_entropy = atof(optarg);
      break;
    case OPT_MIN_SEED_COVERAGE:
      g_min_seed_coverage = atoi(optarg);
      break;
    case OPT_MIN_SEED_ENTROY:
      g_min_seed_entropy = atof(optarg);
      break;
    case OPT_MIN_JUNCTION_COVERAGE:
      g_min_junction_coverage = atoi(optarg);
      break;
    case OPT_MIN_RATIO_NON_ERROR:
      g_min_ratio_non_error = atof(optarg);
      break;
    case OPT_MIN_EXON_LENGTH:
      g_min_exon_length= atoi(optarg);
      break;
    case OPT_MIN_KMERS_PER_GRAPH:
      g_min_kmers_per_graph = atoi(optarg);
      break;
    case OPT_FR_STRAND:
      g_fr_strand = atoi(optarg);
      break;
    case OPT_PAIR_GAP_LENGTH:
      g_pair_gap_length = atoi(optarg);
      break;
    case OPT_DEBUG:
      g_debug = true;
      break;
    default:
      std::cout << usage();
      exit(1);
    }
  } while(next_option != -1);

  if (g_help) {
    std::cout << usage() ;
    exit(1);
  }

  if (reads_file == "") {
    std::cerr << "Error : --input option needs an argument!! " << std::endl;
    std::cout << usage() ;
    exit(1);
  }

  if (g_kmer_length > 32) {
    errAbort(const_cast<char *>("Length of kmer can not be excess 32!\n"));
  }

  if (g_min_anchor_length >= g_kmer_length) {
    g_min_anchor_length = g_kmer_length - 2;
  } 

  if (g_kmer_length != 25) {
    g_min_anchor_length = g_kmer_length - 3;
  } 
 
  if (g_fr_strand != 1 && g_fr_strand != 2) {
    errAbort(const_cast<char *>("--g_fr_strand can only be 1 or 2!\n"));
  }

  return 0;
}


std::string usage () {

  std::stringstream usage_info;
  usage_info
    << std::endl
    << "===============================================================================" << std::endl
    << " Usage: Assemble [--reads/--kmers] <filename>  [opts] " << std::endl
    << "===============================================================================" << std::endl
    << " **Required :" << std::endl
    << " --reads/-i <string>           " << ": the name of the file containing reads" << std::endl;
  //usage_info
  //  << " or " << std::endl
  //  << " --kmers <string>	       " << ": the name of the file containing kmers" << std::endl;
  usage_info
    << std::endl
    << " ** Optional :" << std::endl
    << " --kmer_length/-k <int>        " << ": length of kmer, default: 25." << std::endl
    << " --double_stranded_mode	       " << ": set it true if double stranded mode." << std::endl
    << " --fr_strand<int>	       " << ": strand specific protocol, default: 1 " << std::endl
    << "                                       ( 1 : fr-firststrand, e.g. dUTP, NSR, NNSR " << std::endl
    << "                                         2 : fr-secondstrand, e.g. Strandard SOLID ) " << std::endl
    << " --paired_end		       " << ": set it true if paired reads." << std::endl
    << " --min_seed_coverage <int>     " << ": minimum coverage of seed kmer, default: 2." << std::endl
    << " --min_seed_entropy <float>    " << ": minimum entropy of seed kmer, default: 1.5" << std::endl
    << " --min_kmer_coverage <int>     " << ": minimum coverage of kmer used to extend, default: 1." << std::endl
    << " --min_kmer_entropy <float>    " << ": minimum entroy of kmer used to extend, default: 0.0" << std::endl
    << " --min_junction_coverage <int> " << ": minimum of the coverage of a junction, default: 2." << std::endl
    << " --min_ratio_non_error <float> " << ": min ratio for low/high alternative extension that is "<< std::endl 
    << "                                        not an error, default: 0.05." << std::endl
    << " --pair_gap_length	       " << ": gap length of paired reads, default: 200." << std::endl
    << " --out_dir/-o <string>         " << ": name of directory for output, default : ./RawGraphs " << std::endl
    << " --help/-h                     " << ": display the help information."<< std::endl
    << std::endl;
  usage_info
    << "================================================================================" << std::endl
    << std::endl;

  return (usage_info.str());
}



int main(int argc, char* argv[]) {

  // process command line arguments
  int parse_ret = parse_options(argc, argv);
  if (parse_ret)
    return parse_ret;

  time_t begin = time(NULL);

  if (out_dir != "") {
    //int retcode = mkpath(out_dir.c_str(), 0777);
    bool retcode = boost::filesystem::create_directories(out_dir);
    //if (retcode == -1) {
    if (retcode) {
      if (errno != EEXIST) {
	errAbort(const_cast<char *>("Cannot create directory %s !\n"), out_dir.c_str());
      }
    }
  }
  
  // load data
  std::vector<std::string> data;
  if (reads_file != "")
    load_data(data, reads_file);
  
  KmerMap<info_with_read_set_t> kmer_map(g_kmer_length, g_double_stranded_mode);
  
  assembler(kmer_map, data);

  time_t end = time(NULL);
  std::cerr << "\nDone. " << rg_index << " splcicing grphs are built. Elapsed time: " << (end-begin) <<" s"<< std::endl;  // calculate used time

  data.clear();
  std::vector<std::string> s;
  swap(data,s);
  time_t end1 = time(NULL);
  std::cerr << "Clear memory. Elapsed time: " << (end1-end) <<" s"<< std::endl;
 
  return 0;
}

